### PR TITLE
adding Model.all_valid() and moving graphs

### PR DESCRIPTION
**Pull request by vsoch**
This is a PR to address two issues:

 - #93: we want to move graph templates to be alongside their corresponding model to be easier to find
 - #97: each model should have an .all_valid() function to return a custom queryset.

@abarysh I made this PR smaller we probably will need to have discussion about each of the all_valid() functions, and then to look around and see where else (other than the search.py modules) we would want to use them. I also added one for every model, which might be redundant, but I thought it would be good to have a consistent interface even if it's just Model.all().

Signed-off-by: vsoch <vsochat@stanford.edu>

### CODE REVIEW

**Comment by vsoch** [comment]
Deployed to staging.

File: yeastphenome/apps/conditions/models.py [hunk 1]
```diff
@@ -18,6 +18,18 @@ class Tag(models.Model):
...
+    @classmethod
+    def all_valid(cls):
+        return (
+            cls.objects.annotate(
+                number_of_conditions=Count(
+                    "condition__conditionset__conditions__type__name",
+                )
+            )
+            .filter(number_of_conditions__gte=1)
+            .order_by("-number_of_conditions")
```
Comment (abarysh) [inline yeastphenome/apps/conditions/models.py]: I wonder if we need the `order_by` here -- this method will be called in many contexts, and most of them will need a different sort order. Since `order_by` is not a free operations, should we leave it out and apply as needed?

File: yeastphenome/apps/conditions/models.py [hunk 2]
```diff
@@ -18,6 +18,18 @@ class Tag(models.Model):
     def __str__(self):
         return self.name
 
+    @classmethod
+    def all_valid(cls):
+        return (
+            cls.objects.annotate(
+                number_of_conditions=Count(
```
Comment (abarysh) [inline yeastphenome/apps/conditions/models.py]: Not sure this `Count()` makes sense -- the `Tag` model is linked from both `Conditiontype` and `Condition`. Here you're going `Tag` -> `Condition` -> `Conditionset` -> `Condition` (again) -> `Conditiontype` -> `name`.
I suggest you count number of conditions and number of conditiontypes separately:

```
cls.objects
.annotate(
   number_of_conditions=Count("condition"))
.annotate(
   number_of_conditiontypes=Count("conditiontype"))
.filter(Q(number_of_conditions__gte=1) | Q(number_of_conditiontypes__gte=1))
```

Comment (abarysh) [reply yeastphenome/apps/conditions/models.py]: Actually, now that I think about it, this is also an inefficient way to do it. From what I remember, here we just want to get a list of Tags that have at least 1 condition or 1 conditiontype. The actual number of conditions or conditiontypes can be calculated where it is appropriate, but in the most generic context, the number is not needed. So the best way to do this might be:

```
cls.objects.filter(Q(condition_set__isnull=False) | Q(conditiontype_set__isnull=False) 
```

File: yeastphenome/apps/conditions/models.py [hunk 3]
```diff
@@ -58,6 +70,26 @@ class ConditionType(models.Model):
...
+        return (
+            cls.objects.annotate(
+                number_of_datasets=Count(
+                    "condition__conditionset__dataset",
+                    filter=~Q(
+                        condition__conditionset__dataset__paper__latest_data_status__status__name="not relevant"
+                    ),
+                )
+            )
+            .annotate(
```
Comment (abarysh) [inline yeastphenome/apps/conditions/models.py]: Same here -- I don't think we should pre-compute number_of_papers, if we don't need it to define the set of valid objects here. Since these operations are not free, we should only add them when needed and keep this function as generic as possible.

File: yeastphenome/apps/conditions/models.py [hunk 4]
```diff
@@ -58,6 +70,26 @@ class ConditionType(models.Model):
...
+                number_of_papers=Count(
+                    "condition__conditionset__dataset__paper", distinct=True
+                )
+            )
+            .filter(number_of_datasets__gte=1)
```
Comment (abarysh) [inline yeastphenome/apps/conditions/models.py]: Here I agree, I don't see any other way to do this filtering (i.e., include only conditiontypes that have at least 1 relevant dataset).

File: yeastphenome/apps/conditions/models.py [hunk 5]
```diff
@@ -58,6 +70,26 @@ class ConditionType(models.Model):
...
+            .order_by("-number_of_datasets")
```
Comment (abarysh) [inline yeastphenome/apps/conditions/models.py]: But I would remove the ordering.

File: yeastphenome/apps/conditions/models.py [hunk 6]
```diff
@@ -160,6 +192,10 @@ class Condition(models.Model):
     modified_on = models.DateField(auto_now=True, null=True)
     tags = models.ManyToManyField(Tag, blank=True)
 
+    @classmethod
+    def all_valid(cls):
+        return cls.objects.all()
```
Comment (abarysh) [inline yeastphenome/apps/conditions/models.py]: Here it should be similar to `Conditiontype`:

```
cls.objects.annotate(
    number_of_datasets=Count(
         "conditionset__dataset",
               filter=~Q(
                        conditionset__dataset__paper__latest_data_status__status__name="not relevant"
                    ),
           )
    )
.filter(number_of_datasets__gte=1)
```

File: yeastphenome/apps/conditions/models.py [hunk 7]
```diff
@@ -215,6 +251,10 @@ def __str__(self):
         else:
             return ""
 
+    @classmethod
+    def all_valid(cls):
+        return cls.objects.all()
```
Comment (abarysh) [inline yeastphenome/apps/conditions/models.py]: Here I would consider valid all Conditionsets that are linked to at least 1 dataset:
```
cls.objects.filter(dataset_set__isnull=False) 
```

File: yeastphenome/apps/conditions/models.py [hunk 8]
```diff
@@ -302,6 +342,10 @@ def __str__(self):
...
```
Comment (abarysh) [inline yeastphenome/apps/conditions/models.py:339]: Here I would consider valid all Mediums that are linked to at least 1 dataset:
```
cls.objects.filter(dataset_set__isnull=False) 
```

File: yeastphenome/apps/conditions/search.py [hunk 9]
```diff
@@ -1,54 +1,18 @@
-from django.db.models import Q, Count
+from django.db.models import Q
 
 from yeastphenome.apps.conditions.models import ConditionType, Tag
 from yeastphenome.apps.common.utils import escape_regex
 import itertools
 
 
-def get_tags():
```
Comment (abarysh) [inline yeastphenome/apps/conditions/search.py:8]: Instead of removing these methods, I would invoke here the corresponding `all_valid()` method and then annotate and sort as appropriate for this particular application.

File: yeastphenome/apps/conditions/search.py [hunk 10]
```diff
@@ -1,54 +1,18 @@
...
-            number_of_conditions=Count(
-                "condition__conditionset__conditions__type__name",
-            )
-        )
-        .filter(number_of_conditions__gte=1)
-        .order_by("-number_of_conditions")
-    )
-
-
-def get_conditiontypes():
```
Comment (abarysh) [inline yeastphenome/apps/conditions/search.py:21]: Same here. Call `all_valid()` to get all valid objects, and then annotate & sort as appropriate for this view/application.

File: yeastphenome/apps/datasets/models.py [hunk 11]
```diff
@@ -16,6 +16,10 @@ class Collection(models.Model):
     ploidy = models.IntegerField(null=True, blank=True)
     description = models.TextField(null=True, blank=True)
 
+    @classmethod
+    def all_valid(cls):
+        return cls.objects.all()
```
Comment (abarysh) [inline yeastphenome/apps/datasets/models.py]: This should be:

```
cls.objects.exclude(paper__latest_data_status__status__name="not relevant")
```

File: yeastphenome/apps/datasets/models.py [hunk 12]
```diff
@@ -48,6 +56,10 @@ def __str__(self):
         else:
             return "%s" % self.sourcetype
 
+    @classmethod
+    def all_valid(cls):
+        return cls.objects.all()
```
Comment (abarysh) [inline yeastphenome/apps/datasets/models.py]: Sources linked to at least 1 relevant dataset:
```
cls.objects.filter(dataset_set__isnull=False).exclude(dataset__paper__latest_data_status__status__name="not relevant")
```

File: yeastphenome/apps/datasets/models.py [hunk 13]
```diff
@@ -95,6 +111,10 @@ class Tag(models.Model):
     name = models.CharField(max_length=200, null=True, blank=True)
```
Comment (abarysh) [inline yeastphenome/apps/datasets/models.py]: I feel like we should have `null=False, blank=False` for the name field. Is it really a tag if we don't have a name for it?

File: yeastphenome/apps/datasets/models.py [hunk 14]
```diff
@@ -95,6 +111,10 @@ class Tag(models.Model):
...
     description = models.TextField(max_length=1000, null=True, blank=True)
 
+    @classmethod
+    def all_valid(cls):
+        return cls.objects.exclude(name__isnull=True)
```
Comment (abarysh) [inline yeastphenome/apps/datasets/models.py]: The `null=False, blank=False` in the name field should eliminate the need to specify `.exclude(name__isnull=True)`. But similar to the Condition and Conditiontype `Tag`, we should have:
```
cls.objects.filter(dataset_set__isnull=False)
```

File: yeastphenome/apps/datasets/models.py [hunk 15]
```diff
@@ -211,7 +231,7 @@ def get_absolute_url(self):
         return reverse("datasets:detail", args=[self.id])
 
     @classmethod
-    def all(cls):
+    def all_valid(cls):
         return cls.objects.exclude(
             Q(paper__latest_data_status__status__name__exact="not relevant")
             | Q(paper__data_statuses__name__exact="not relevant")
```
Comment (abarysh) [inline yeastphenome/apps/datasets/models.py]: Only the 1st condition is needed, right? "not relevant" is always the latest status (after a paper is set to "not relevant" no other statuses are added).

File: yeastphenome/apps/genes/models.py [hunk 16]
```diff
@@ -11,6 +11,10 @@ class GeneAlias(models.Model):
 
     name = models.CharField(max_length=250, null=True, blank=True, unique=True)
```
Comment (abarysh) [inline yeastphenome/apps/genes/models.py:12]: Similar to above: `name` is the only field for `GeneAlias` and it should never be null or blank.

File: yeastphenome/apps/genes/models.py [hunk 17]
```diff
@@ -11,6 +11,10 @@ class GeneAlias(models.Model):
...
 
+    @classmethod
+    def all_valid(cls):
+        return cls.objects.exclude(name__isnull=True)
```
Comment (abarysh) [inline yeastphenome/apps/genes/models.py:16]: Setting name to `null=False, blank=False` will eliminate the need to specify `.exclude(name__isnull=True)`

File: yeastphenome/apps/papers/models.py [hunk 18]
```diff
@@ -18,6 +18,10 @@ class Status(models.Model):
     description = models.TextField(blank=True, null=True)
     is_valid = models.BooleanField()
```
Comment (abarysh) [inline yeastphenome/apps/papers/models.py:19]: This field has now become a misnomer. `is_valid=True` was (so far) used to indicate statuses that correspond to a situation where we already have or expect to have data for a paper (e.g., "loaded", "to request", "request sent" etc.). In contrast `is_valid=False` corresponds to statuses where, for various reasons, the data will not be available (e.g., "not available", "request abandoned", "not relevant"). 
"Not relevant" here is an outlier -- it indicates a paper that exists in the DB, but should always be ignored because it is outside of our project's scope and we are only keeping it for our records. In contrast, "not available" means that the paper is relevant, we should have gotten data, but the authors confirmed that the data has been lost forever. "Request abandoned" means that we failed to get a response from the authors and, as far as we're concerned, the data is lost forever.
So, long story short -- at some point we should rename `is_valid` to something else. But for now, it shouldn't be used in the `all_valid()` method (we want all statuses there).

File: yeastphenome/apps/papers/models.py [hunk 19]
```diff
@@ -18,6 +18,10 @@ class Status(models.Model):
...
 
+    @classmethod
+    def all_valid(cls):
+        return cls.objects.exclude(is_valid=False).all()
```
Comment (abarysh) [inline yeastphenome/apps/papers/models.py]: Should just be
```
cls.objects.all()
```

File: yeastphenome/apps/papers/models.py [hunk 20]
```diff
@@ -59,7 +63,7 @@ class Paper(models.Model):
     )
 
     @classmethod
-    def all(cls):
+    def all_valid(cls):
         return cls.objects.exclude(
             Q(latest_data_status__status__name__exact="not relevant")
             | Q(data_statuses__name__exact="not relevant")
```
Comment (abarysh) [inline yeastphenome/apps/papers/models.py]: The second Q condition is not needed.

File: yeastphenome/apps/phenotypes/models.py [hunk 21]
```diff
@@ -11,6 +11,10 @@ class Tag(models.Model):
     def __str__(self):
         return self.name
 
+    @classmethod
+    def all_valid(cls):
+        return cls.objects.all()
```
Comment (abarysh) [inline yeastphenome/apps/phenotypes/models.py]: For consistency, you might want too all tags associated with at least 1 observable:
```
cls.objects.filter(observable_set__isnull=False)
```

File: yeastphenome/apps/phenotypes/models.py [hunk 22]
```diff
@@ -49,6 +53,10 @@ class Observable(models.Model):
     modified_on = models.DateField(auto_now=True, null=True)
     tags = models.ManyToManyField(Tag, blank=True)
 
+    @classmethod
+    def all_valid(cls):
+        return cls.objects.all()
```
Comment (abarysh) [inline yeastphenome/apps/phenotypes/models.py]: Here I would keep all observables that are linked to at least 1 phenotype:
```
cls.objects.filter(phenotype_set__isnull=False)
```

File: yeastphenome/apps/phenotypes/models.py [hunk 23]
```diff
@@ -204,6 +216,10 @@ def papers(self):
             .distinct()
         )
 
+    @classmethod
+    def all_valid(cls):
+        return cls.papers_all()
```
Comment (abarysh) [inline yeastphenome/apps/phenotypes/models.py]: Hmm.. this looks like a bug (as written, it will give you a set of papers, instead of phenotypes). Shouldn't it be:
```
cls.objects.filter(dataset_set__isnull=False)
```

**Comment by abarysh** [comment]
I went through all of the `all_valid()` methods and made multiple comments. I think it was a great idea to have this method, but it might takes us some time to fine-tune the definition of valid for each model. I hope my comments are clear and helpful.

**Comment by vsoch** [comment]
Your comments are great! I'll go through them carefully tomorrow and make all the changes.

File: yeastphenome/apps/conditions/models.py (hunk 1 above)
Comment (vsoch) [reply yeastphenome/apps/conditions/models.py]: Removed `order_by`!  I don't think we use this one anywhere on the site.

File: yeastphenome/apps/conditions/models.py (hunk 2 above)
Comment (vsoch) [reply yeastphenome/apps/conditions/models.py]: This wound up working:

```
    @classmethod
    def all_valid(cls):
        return cls.objects.filter(Q(condition__isnull=False) | Q(conditiontype__isnull=False))
```

File: yeastphenome/apps/conditions/models.py (hunk 3 above)
Comment (vsoch) [reply yeastphenome/apps/conditions/models.py]: Removed, and added to the browse view (where i believe we used it along with the sorting).

File: yeastphenome/apps/conditions/models.py (hunk 7 above)
Comment (vsoch) [reply yeastphenome/apps/conditions/models.py]: Used `cls.objects.filter(dataset__isnull=False) `

File: yeastphenome/apps/conditions/models.py (hunk 8 above)
Comment (vsoch) [reply yeastphenome/apps/conditions/models.py:339]: Here I used `cls.objects.filter(dataset__isnull=False) `

File: yeastphenome/apps/conditions/search.py (hunk 9 above)
Comment (vsoch) [reply yeastphenome/apps/conditions/search.py:8]: This function doesn't appear to be used anywhere (the get_tags) so I think it's okay to remove, per #88

File: yeastphenome/apps/conditions/search.py (hunk 10 above)
Comment (vsoch) [reply yeastphenome/apps/conditions/search.py:21]: The only annotation / sort that was needed was for the condition tag browser, so I added it there to the function in views.py, using ConditionType.all_valid(). We don't have this function convention for other models (e.g., get_modelnames) so I want to suggest we keep what we currently have and not add this function back (I'll show you when I push changes).

File: yeastphenome/apps/datasets/models.py (hunk 11 above)
Comment (vsoch) [reply yeastphenome/apps/datasets/models.py]: Gotcha! Wound up using `cls.objects.exclude(dataset__paper__latest_data_status__status__name="not relevant")  `

File: yeastphenome/apps/datasets/models.py (hunk 12 above)
Comment (vsoch) [reply yeastphenome/apps/datasets/models.py]: Looks like this wound up working:
```
cls.objects.filter(data_source__isnull=False).exclude(data_source__paper__latest_data_status__status__name="not relevant") 
```

File: yeastphenome/apps/datasets/models.py (hunk 13 above)
Comment (vsoch) [reply yeastphenome/apps/datasets/models.py]: Agree! This caught my eye too. We certainly don't have any empty tags in the database so this should be okay to add.

File: yeastphenome/apps/datasets/models.py (hunk 14 above)
Comment (vsoch) [reply yeastphenome/apps/datasets/models.py]: Adjusted to `cls.objects.filter(dataset__isnull=False) `

File: yeastphenome/apps/datasets/models.py (hunk 15 above)
Comment (vsoch) [reply yeastphenome/apps/datasets/models.py]: Yep just checked, removing the last two returns the same number of datasets.

File: yeastphenome/apps/genes/models.py (hunk 16 above)
Comment (vsoch) [reply yeastphenome/apps/genes/models.py:12]: Definitely, but I think we should hold off any kind of migration in the genes model, since it would likely show up again under datastes. This is a small change and I'll add a note to do it when we rename the table.

File: yeastphenome/apps/papers/models.py (hunk 19 above)
Comment (vsoch) [reply yeastphenome/apps/papers/models.py]: Fixed.

File: yeastphenome/apps/papers/models.py (hunk 20 above)
Comment (vsoch) [reply yeastphenome/apps/papers/models.py]: And from before, it looks like we can remove the third too (yields same number of results).

File: yeastphenome/apps/phenotypes/models.py (hunk 21 above)
Comment (vsoch) [reply yeastphenome/apps/phenotypes/models.py]: Used `cls.objects.filter(observable__isnull=False)`

File: yeastphenome/apps/phenotypes/models.py (hunk 22 above)
Comment (vsoch) [reply yeastphenome/apps/phenotypes/models.py]: Used `cls.objects.filter(phenotype__isnull=False)`

File: yeastphenome/apps/phenotypes/models.py (hunk 23 above)
Comment (vsoch) [reply yeastphenome/apps/phenotypes/models.py]: This was definitely me not thinking! It's now `cls.objects.filter(dataset__isnull=False)`

**Comment by vsoch** [comment]
Deployed to staging! And strange, I had expected changing the null / empty fields for the Tag to require a migration, but I don't see that one is needed.

**Comment by abarysh** [comment]
I think it's ok to merge into master & deploy to production.

**Comment by vsoch** [comment]
Deployed to production. @abarysh did you want to do the table rename today, and if so, how should we go about it?

**Comment by abarysh** [comment]
Yes, I'm about to start now. Here's my plan:
- create a copy of the datasets_gene table and related
- rename the copies to genes_gene etc.
- make the appropriate change to the code and test locally
- if everything works well, submit a PR with the change
- in a week from now (after some more extensive testing), delete the old tables

How does it sound?

**Comment by vsoch** [comment]
That sounds good! For the third bullet, you'd basically remove the Meta value that says that gene is still under the datasets module. And given that the table will exist, my fingers are crossed that it doesn't detect any migrations.
